toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -185,13 +313,206 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
+		return false, err
+	}
+	if lookup == nil || lookup.RunNode == nil {
+		err = fmt.Errorf("can't find run lookup '%v' in graph", lookup)
+		return false, err
+	}
+
+	if lookup.BuildNode != realBuildNode {
+		err = fmt.Errorf("real build node '%v' does not match build node in the graph lookup '%v'", realBuildNode, lookup.BuildNode)
+		return false, err
+	}
+
+	realRunNode := lookup.RunNode
+	// We need to copy the node since the fetcher may on occasion choose a different package to provide something than
+	// we want. We only consider a delta node valid if the .rpm matches exactly. If we tryy and fail to download a delta
+	// the node would be left in a bad state.
+	nodeCopy := realBuildNode
+
+	// Get the final output path for the build node if we don't convert it to a delta node
+	originalRpmPath := realBuildNode.RpmPath
+	foundFinalRPM, _ := file.IsFile(originalRpmPath)
+
+	// Only download dependencies for delta RPMs if we don't already have the RPM in the out/RPMS folder
+	if !foundFinalRPM {
+		// Flag the node as a delta node now so the resolver knows this isn't a critical issue if it fails to resolve
+		nodeCopy.State = pkggraph.StateDelta
+		resolveErr := resolveSingleNode(cloner, nodeCopy, downloadDependencies, nil, make(map[string]bool), make(map[string]bool), deltaRpmDir)
+		// Failing to clone a dependency should not halt a build.
+		// The build should continue and attempt best effort to build as many packages as possible.
+		if resolveErr != nil {
+			logger.Log.Warnf("Can't find delta RPM to download for %s-%s: %s", nodeCopy.VersionedPkg.Name, nodeCopy.VersionedPkg.Version, resolveErr)
+			return false, nil
 		}
+		logger.Log.Tracef("Updating real node '%s' with info from newly cached delta node '%s'", realBuildNode, nodeCopy)
+
+		// Check that the RPM we are getting matches the expected out/RPM path using the base name of the file path
+		cachedRPMPath := nodeCopy.RpmPath
+		if filepath.Base(cachedRPMPath) != filepath.Base(originalRpmPath) {
+			logger.Log.Warnf("cached delta RPM '%s' does not match expected RPM '%s', skipping", filepath.Base(cachedRPMPath), filepath.Base(originalRpmPath))
+			return false, nil
+		} else {
+			logger.Log.Infof("Delta RPM found for '%s-%s'.", realBuildNode.VersionedPkg.Name, realBuildNode.VersionedPkg.Version)
+		}
+
+		realBuildNode.State = pkggraph.StateDelta
+		realRunNode.State = pkggraph.StateDelta
+
+		// Record the original path to delta path mapping so we can fix up the implicit nodes later.
+		(*pathMap)[originalRpmPath] = cachedRPMPath
+		realRunNode.RpmPath = cachedRPMPath
+		realBuildNode.RpmPath = cachedRPMPath
+
+		logger.Log.Tracef("PathMap updated: %v", *pathMap)
+		logger.Log.Debugf("Converted delta build node is now: '%s'", realBuildNode)
+		logger.Log.Debugf("Converted delta run node is now: '%s'", realRunNode)
+	} else {
+		logger.Log.Infof("Already have a RPM for '%s' at '%s'.", realBuildNode.VersionedPkg.Name, originalRpmPath)
```

Comment:
>logger.Log.Infof("Already have a RPM for '%s' at '%s'.", realBuildNode.VersionedPkg.Name, originalRpmPath) [](http://example.com/codeflow?start=2&length=106)

In the case where we already have the RPM locally, should we still update `realRunNode.RpmPath` or is it already guaranteed to be set at this point?

Reply:
The default for a build node will point to out/RPMS unless we change it here, it is safe to leave it alone.